pkg/search/postgres/storecache.go

```diff
@@ -0,0 +1,329 @@
+		permissionChecker: checker,
+		cache:             make(map[string]PT),
+		useCache:          true,
+		underlyingStore:   underlyingStore,
+	}
+	store.cacheLock.Lock()
+	defer store.cacheLock.Unlock()
+	store.resetCacheNoLock()
+	return store
+}
+
+// CachedStore implements subset of Store interface for resources with single ID.
+type CachedStore[T any, PT unmarshaler[T]] struct {
+	mutex                         sync.RWMutex
+	schema                        *walker.Schema
+	pkGetter                      primaryKeyGetter[T, PT]
+	setCacheOperationDurationTime durationTimeSetter
+	permissionChecker             walker.PermissionChecker
+	targetResource                permissions.ResourceMetadata
+	underlyingStore               Store[T, PT]
+	cache                         map[string]PT
+	useCache                      bool
+	cacheLock                     sync.RWMutex
+}
+
+// Upsert saves the current state of an object in storage.
+func (c *CachedStore[T, PT]) Upsert(ctx context.Context, obj PT) error {
+	dbErr := c.underlyingStore.Upsert(ctx, obj)
+	c.cacheLock.Lock()
+	defer c.cacheLock.Unlock()
+	if dbErr != nil {
+		c.resetCacheNoLock()
+		return dbErr
+	}
+	c.addToCacheNoLock(obj)
+	return nil
+}
+
+// UpsertMany saves the state of multiple objects in the storage.
+func (c *CachedStore[T, PT]) UpsertMany(ctx context.Context, objs []PT) error {
+	dbErr := c.underlyingStore.UpsertMany(ctx, objs)
+	c.cacheLock.Lock()
+	defer c.cacheLock.Unlock()
+	if dbErr != nil {
+		c.resetCacheNoLock()
+		return dbErr
+	}
+	for _, obj := range objs {
+		c.addToCacheNoLock(obj)
+	}
+	return nil
+}
+
+// Delete removes the object associated to the specified ID from the store.
+func (c *CachedStore[T, PT]) Delete(ctx context.Context, id string) error {
+	dbErr := c.underlyingStore.Delete(ctx, id)
+	c.cacheLock.Lock()
+	defer c.cacheLock.Unlock()
+	if dbErr != nil {
+		c.resetCacheNoLock()
+		return dbErr
+	}
+	delete(c.cache, id)
+	return nil
+}
+
+// DeleteMany removes the objects associated to the specified IDs from the store.
+func (c *CachedStore[T, PT]) DeleteMany(ctx context.Context, identifiers []string) error {
+	dbErr := c.underlyingStore.DeleteMany(ctx, identifiers)
+	c.cacheLock.Lock()
+	defer c.cacheLock.RUnlock()
+	if dbErr != nil {
+		c.resetCacheNoLock()
+		return dbErr
+	}
+	for _, id := range identifiers {
+		delete(c.cache, id)
+	}
+	return nil
+}
+
+// Exists tells whether the ID exists in the store.
+func (c *CachedStore[T, PT]) Exists(ctx context.Context, id string) (bool, error) {
+	c.cacheLock.RLock()
+	defer c.cacheLock.RUnlock()
+	obj, found := c.cache[id]
+	if !found {
+		return false, nil
+	}
+	return c.isReadAllowed(ctx, obj), nil
+}
+
+// Count returns the number of objects in the store.
+func (c *CachedStore[T, PT]) Count(ctx context.Context) (int, error) {
+	c.cacheLock.RLock()
+	defer c.cacheLock.RUnlock()
+	count := 0
+	walkErr := c.walkCacheNoLock(ctx, func(obj PT) error {
+		count++
+		return nil
+	})
```

Comment:
Hmm, assuming the map is handled properly (i.e. cleanup is done correctly), this could simply be a `return len(c.cache)` instead of iterating over the map contents?

Reply:
Need to iterate over the map to apply sac properly

Reply:
Indeed, `walkCacheNoLock` filters out items excluded by SAC checks.